lib/java/persistance/src/main/java/org/enso/persist/PerGenerator.java

```diff
@@ -106,6 +114,63 @@ final int versionStamp() {
     return map.versionStamp;
   }
 
+  private int registerReference(Persistance.Reference<?> ref) {
+    var obj = ref.get(Object.class);
+    var existingId = pendingReferences.get(obj);
+    if (existingId == null) {
+      var currentSize = countReferences++;
+      pendingReferences.put(obj, currentSize);
+      return currentSize;
+    } else {
+      return existingId;
+    }
+  }
+
+  /**
+   * Writes an object into the buffer. Writes also all {@link Persistance.Reference} that were left
+   * pending during the serialization.
+   *
+   * @param obj the object to write down
+   * @return location of the table {@code int size and then int[size]}
+   */
+  private int writeObjectAndReferences(Object obj) throws IOException {
+    var objAt = writeObject(obj);
+
+    var refsOut = new ByteArrayOutputStream();
+    var refsData = new DataOutputStream(refsOut);
+    refsData.writeInt(-1); // space for size of references
+    refsData.writeInt(objAt); // the main object
+    var count = 1;
+    while (!pendingReferences.isEmpty()) {
+      var round = new ArrayList<>(pendingReferences.entrySet());
+      round.sort(
+          (e1, e2) -> {
+            return e1.getValue() - e2.getValue();
+          });
+      pendingReferences.clear();
+      for (var entry : round) {
+        count++;
+        if (obj == entry.getKey()) {
+          refsData.writeInt(-1);
+          continue;
+        }
+        var at = writeObject(entry.getKey());
+        assert count == entry.getValue() : "Expecting " + count + " got " + entry.getValue();
+        refsData.writeInt(at);
```

Comment:
```suggestion
        if (obj == entry.getKey()) {
          refsData.writeInt(-1);
          continue;
        } else {
          var at = writeObject(entry.getKey());
          assert count == entry.getValue() : "Expecting " + count + " got " + entry.getValue();
          refsData.writeInt(at);
        }
```

This is a very subjective suggestion, but IMHO in this case an `if-else` is much easier to follow than an `if+continue`. It is clearer that this is a 2-way branch from the overall structure - whereas currently at first glance it is unclear if the code after `if` may also be executed if the `if` was hit (it's not because of the `continue`, but that's not visible at first glance).

I'm sorry if I'm too annoying with such suggestions, but the clearer the intent of the code is, the easier it is to understand it as the mental load needed to understand the overall structure decreases, freeing more "brain RAM" to actually think about the underlying concepts. For super easy code that may not matter too much, but this one is not immediately obvious, so making it easier to understand would is IMO beneficial.

Reply:
Anyway, the above is just a nitpick.

My major problem is - I don't think I understand why adding the `-1` here works. I understand we add -1 because we don't yet know the offset for `obj`. But when do we 'come back' and fix-up these -1s to actual proper ids?

Reply:
Oh, or is THAT related to the code I asked about moments before - the `InputCache`'s `if (refs[i] < 0) { refs[i] = refs[0]; }` rewrite? I guess most likely YES!

I really think some explanation is needed here. I understood this correlation after carefully reading the whole code and trying to deeply understand the serialization process - but at first look it seemed like just some magic numbers. It would be good if understanding this code did not require guessing.

Reply:
Also a style thing, so feel free to ignore, but I think it makes more sense to do this `-1` reference fix-up at write time.

While performance-wise it probably does not matter at all, it just seems (to me) to make more sense 'philosophically' for the writer to handle the looping complexity - the reader IMHO should be as dumb as possible, just reading offsets and recreating the necessary structure.

The benefit of that is that the _whole_ logic for the reference reconstruction would be in one module, one place, making it easier to understand what was meant here (although I still think there should be some comment describing the idea). But when it's split over multiple modules like now, and there's no comments, it's much harder to find out what this logic is all about. I'm still lucky that I'm reviewing a single PR so I just see in the diff the relevant changes - if I was reading these modules in the context of the _whole_ codebase, deciphering what this algorithm scattered through multiple files is meant to do would be even harder.

Reply:
Yes Radek, you "dissassembled" the idea correctly. Since a958804348 it shall be clearer that these two places are connected.